File: C/Chapter 1 - Arrays and Strings/Q1.c
/*
    Is Unique: Implement an algorithm to determine if a string has all unique characters.

               What if you can't use additional data structures?
*/

//Assumption - Only Ascii Strings
#include <string.h>
#include <stdio.h>
int isUnique(char* str);
int main(){
    char* string = "This is a Test String";
    char* passstring = "qwertyu";
    
    if(isUnique(string)){
        printf("\"%s\" is unique\n", string);
    } else {
        printf("\"%s\" is not unique\n", string);
    }

    if(isUnique(passstring)){
        printf("\"%s\" is unique\n", passstring);
    } else {
        printf("\"%s\" is not unique\n", passstring);
    }
return 0;
}

int isUnique(char* str){
int flagArray[126] = {0}; //assuming normal ascii
int i;
for(i = 0; i < strlen(str); i++){
    flagArray[str[i]]++;
    if(flagArray[str[i]] > 1){
        return 0;
    }
}

return 1;
}

File: Java/Chapter 1 - Arrays and Strings/Q1.java
/**
* Q1) Create a function to determine if a string has any duplicate characters in it
*/
public class Q1 {
    public static void main(String[] args){
        test_fn();
    }
    /**
    * This function will test if a string has any duplicate characters.
    * It is only valid for the E-Ascii or Ascii character sets    
    * 
    * @param str the string to be tested 
    * @return true if the string has only unique characters, false if there are any duplicates
    */
    public static boolean is_unique(String str){
        int[] flags = new int[256]; // E-ascii

        //toUpperCase because T and t should still be counted as the same character
        for (char c : str.toUpperCase().toCharArray()){
            flags[c]++;
            if  (flags[c] > 1) {
                return false;
            }
        }
        return true;
    }
    /**
    *    Tests the is_unique function against 11 test cases and prints to terminal if the test is successful
    */
    public static void test_fn(){
        String[] testData = {"test", "tes", "perl", "TesT", "abcdefg", "abcddf", "abbbbcd", "AbCdEfG", "123aaa", "ab11ab", "{}}{"};
        boolean[] testRes = {false, true, true, false, true, false, false, true, false, false, false};
        for(int i = 0; i < testData.length; i++) {
            if(is_unique(testData[i]) != testRes[i]){
                System.out.println("Testing of function failed");
                return;
            }
        }
        System.out.println("Testing successful.");
        return;
    }
}
